lib/src/rules/lines_longer_than_80_chars.dart

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) 2018, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:linter/src/analyzer.dart';
+
+const _desc = r'AVOID lines longer than 80 characters.';
+
+const _details = r'''
+
+**AVOID** lines longer than 80 characters
+
+Readability studies show that long lines of text are harder to read because your
+eye has to travel farther when moving to the beginning of the next line. This is
+why newspapers and magazines use multiple columns of text.
+
+If you really find yourself wanting lines longer than 80 characters, our
+experience is that your code is likely too verbose and could be a little more
+compact. The main offender is usually `VeryLongCamelCaseClassNames`. Ask
+yourself, “Does each word in that type name tell me something critical or
+prevent a name collision?” If not, consider omitting it.
+
+Note that dartfmt does 99% of this for you, but the last 1% is you. It does not
+split long string literals to fit in 80 columns, so you have to do that
+manually.
+
+We make an exception for URIs and file paths. When those occur in comments or
+strings (usually in imports and exports), they may remain on a single line even
+if they go over the line limit. This makes it easier to search source files for
+a given path.
+''';
+
+class LinesLongerThan80Chars extends LintRule implements NodeLintRule {
+  LinesLongerThan80Chars()
+      : super(
+            name: 'lines_longer_than_80_chars',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry) {
+    final visitor = new _Visitor(this);
+    registry.addCompilationUnit(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor {
+  final LintRule rule;
+
+  _Visitor(this.rule);
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    final lineInfo = node.lineInfo;
+
+    final allowedLineVisitor = new _AllowedLongLineVisitor(lineInfo);
+    node.accept(allowedLineVisitor);
+    final allowedCommentVisitor = new _AllowedCommentVisitor(lineInfo);
+    node.accept(allowedCommentVisitor);
+
+    final allowedLines = []
+      ..addAll(allowedLineVisitor.allowedLines)
+      ..addAll(allowedCommentVisitor.allowedLines);
+
+    final lineCount = lineInfo.lineCount;
+    for (int i = 0; i < lineCount; i++) {
```

Comment:
It occurs to me that we could significantly improve performance if we moved most of this loop to the beginning of the method. Instead of checking `allowedLines.contains(i + 1)` and reporting a lint, we could add the index of the long line to a collection of long lines and exit early if the collection was empty. We'd need another loop iterating over the collection of long lines to see whether they were allowed.

This would skip two visits of the AST for files that don't contain long lines, which is likely to be most of them for any package that has this rule enabled.

Reply:
done